compiler/ir/backend.wasm/src/org/jetbrains/kotlin/backend/wasm/ir2wasm/DeclarationGenerator.kt

```diff
@@ -0,0 +1,329 @@
+
+        // Generate function type
+        val watName = declaration.fqNameWhenAvailable.toString()
+        val irParameters = declaration.getEffectiveValueParameters()
+        val wasmFunctionType =
+            WasmFunctionType(
+                name = watName,
+                parameterTypes = irParameters.map {
+                    val t = context.transformValueParameterType(it)
+                    if (importedName != null && t is WasmRefNullType) {
+                        WasmEqRef
+                    } else {
+                        t
+                    }
+                },
+                resultTypes = listOfNotNull(
+                    context.transformResultType(declaration.returnType).let {
+                        if (importedName != null && it is WasmRefNullType) WasmEqRef else it
+                    }
+                )
+            )
+        context.defineFunctionType(declaration.symbol, wasmFunctionType)
+
+        val isIntrinsic = declaration.hasWasmReinterpretAnnotation() || declaration.getWasmOpAnnotation() != null
+
+        if (declaration is IrSimpleFunction) {
+            if (declaration.modality == Modality.ABSTRACT) return
+            if (declaration.isFakeOverride) return
+
+            val isVirtual = declaration.isOverridableOrOverrides
+
+            if (isVirtual) {
+                // Register function as virtual, meaning this function
+                // will be stored Wasm table and could be called indirectly.
+                context.registerVirtualFunction(declaration.symbol)
+            }
+
+            if (!isVirtual && isIntrinsic) {
+                // Calls to non-virtual intrinsic functions are replaced with something else.
+                // No need to generate them.
+                return
+            }
+        }
+
+        assert(declaration == declaration.realOverrideTarget) {
+            "Sanity check that $declaration is a real function that can be used in calls"
+        }
+
+        if (importedName != null) {
+            // Imported functions don't have bodies. Declaring the signature:
+            context.defineFunction(
+                declaration.symbol,
+                WasmFunction.Imported(watName, wasmFunctionType, importedName)
+            )
+            // TODO: Support re-export of imported functions.
+            return
+        }
+
+        val function = WasmFunction.Defined(watName, wasmFunctionType)
+        val functionCodegenContext = WasmFunctionCodegenContextImpl(
+            declaration,
+            function,
+            backendContext,
+            context
+        )
+
+        for (irParameter in irParameters) {
+            functionCodegenContext.defineLocal(irParameter.symbol)
+        }
+
+        val exprGen = functionCodegenContext.bodyGen
+        val bodyBuilder = BodyGenerator(functionCodegenContext)
+
+        if (isIntrinsic) {
+            bodyBuilder.tryToGenerateWasmOpIntrinsicCall(declaration)
```

Comment:
I doubt it works, at least we have to load arguments

Reply:
Yes. We have `if (isIntrinsic) return` earlier :)